libquery/src/AttributedGraph.cpp

```
@@ -363,130 +373,324 @@ size_t AttributedGraph::matchCypherQuery(const char* cypherQueryStr) {
+  // TODO equivalent for edges and paths
+  // prefix sum of edges for nodes
   std::vector<size_t> prefixSum;
   std::vector<std::pair<size_t, size_t>> starEdgeList;
   std::vector<QueryEdgeData> starEdgeData;
-  galois::StatTimer compileTime("IRCompileTime");
 
+  galois::StatTimer compileTime("IRCompileTime");
   compileTime.start();
 
-  // loop through the nodes only if no edges exist
-  if (queryEdges.size() == 0) {
+  //////////////////////////////////////////////////////////////////////////////
+  // Book keeping of nodes/edges
+  //////////////////////////////////////////////////////////////////////////////
+  if (queryEdges.size() != 0) {
+    // loop through all edges parsed from compiler and do bookkeeping:
+    // - count the number of nodes that will exist
+    // - get labels
+    // - count degrees for each source vertex (by extension, count # edges)
+    // - get star edges if any
+    for (size_t j = 0; j < queryEdges.size(); ++j) {
+      // ids of nodes of this edge
+      // assumes that the id is an int; also assumes that the parser end has
+      // correctly bound the src/dst IDs
+      size_t srcID = std::stoi(queryEdges[j].caused_by.id);
+      size_t dstID = std::stoi(queryEdges[j].acted_on.id);
+
+      // allocate more memory for nodes if node ids go past what we currently
+      // have
+      if (srcID >= numQueryNodes) {
+        numQueryNodes = srcID + 1;
+      }
+      if (dstID >= numQueryNodes) {
+        numQueryNodes = dstID + 1;
+      }
+      nodeTypes.resize(numQueryNodes, NULL);
+      nodeContains.resize(numQueryNodes, "");
+      prefixSum.resize(numQueryNodes, 0);
+      node_index_to_var_name.resize(numQueryNodes, "");
+
+      // node types check: save node types for each id
+      // NOTE: This saves the FIRST type encountered by the source ID
+      if (nodeTypes[srcID] == NULL) {
+        nodeTypes[srcID] = queryEdges[j].caused_by.labels;
+      } else {
+        // assert(std::string(nodeTypes[srcID]) ==
```

?

---

Bad assertion that was checking if the user was constructing a sane query that ended up biting us when the a passed in query wasn't as expected (though correctness wise it doesn't matter as long as the user doesn't redefine a label on a node).

Will remove.

---

Assertion-wise, if this is something on user-defined data, then it would be better to return an explicit error to the user because we can't guarantee what they will submit as a query. In general it would be good to be very defensive around user input.